pkg/controller/autoscaling/elasticsearch/status/status.go

```diff
@@ -0,0 +1,212 @@
+	PolicyStates []PolicyState `json:"state"`
+	// LastModificationTime is the last time the resources have been updated, used by the cooldown algorithm.
+	LastModificationTime metav1.Time `json:"lastModificationTime"`
+}
+
+func (s *Status) GetNamedTierResources(policyName string) (resources.NodeSetsResources, bool) {
+	for _, policyStatus := range s.AutoscalingPolicyStatuses {
+		if policyStatus.Name == policyName {
+			return resources.NodeSetsResources{
+				Name:             policyStatus.Name,
+				NodeSetNodeCount: policyStatus.NodeSetNodeCount,
+				NodeResources:    policyStatus.ResourcesSpecification,
+			}, true
+		}
+	}
+	return resources.NodeSetsResources{}, false
+}
+
+func (s *Status) GetLastModificationTime(policyName string) (metav1.Time, bool) {
+	for _, policyState := range s.AutoscalingPolicyStatuses {
+		if policyState.Name == policyName {
+			return policyState.LastModificationTime, true
+		}
+	}
+	return metav1.Time{}, false
+}
+
+type AutoscalingPolicyStatusBuilder struct {
+	policyName           string
+	namedTierResources   resources.NodeSetsResources
+	lastModificationTime metav1.Time
+	states               map[PolicyStateType]PolicyState
+}
+
+func NewAutoscalingPolicyStatusBuilder(name string) *AutoscalingPolicyStatusBuilder {
+	return &AutoscalingPolicyStatusBuilder{
+		policyName: name,
+		states:     make(map[PolicyStateType]PolicyState),
+	}
+}
+
+func (psb *AutoscalingPolicyStatusBuilder) Build() AutoscalingPolicyStatus {
+	policyStates := make([]PolicyState, len(psb.states))
+	i := 0
+	for _, v := range psb.states {
+		policyStates[i] = PolicyState{
+			Type:     v.Type,
+			Messages: v.Messages,
+		}
+		i++
+	}
+	return AutoscalingPolicyStatus{
+		Name:                   psb.policyName,
+		NodeSetNodeCount:       psb.namedTierResources.NodeSetNodeCount,
+		ResourcesSpecification: psb.namedTierResources.NodeResources,
+		LastModificationTime:   psb.lastModificationTime,
+		PolicyStates:           policyStates,
+	}
+}
+
+// SetNamedTierResources sets the compute resources associated to a tier.
+func (psb *AutoscalingPolicyStatusBuilder) SetNamedTierResources(namedTierResources resources.NodeSetsResources) *AutoscalingPolicyStatusBuilder {
+	psb.namedTierResources = namedTierResources
+	return psb
+}
+
+func (psb *AutoscalingPolicyStatusBuilder) SetLastModificationTime(lastModificationTime metav1.Time) *AutoscalingPolicyStatusBuilder {
+	psb.lastModificationTime = lastModificationTime
+	return psb
+}
+
+// WithEvent records a new event (type + message) for the tier.
+func (psb *AutoscalingPolicyStatusBuilder) WithEvent(stateType PolicyStateType, message string) *AutoscalingPolicyStatusBuilder {
+	if policyState, ok := psb.states[stateType]; ok {
+		policyState.Messages = append(policyState.Messages, message)
+		psb.states[stateType] = policyState
+		return psb
+	}
+	psb.states[stateType] = PolicyState{
+		Type:     stateType,
+		Messages: []string{message},
+	}
+	return psb
+}
+
+type PolicyStateType string
```

Comment:
This is something I didn't try to optimize yet. The API has been designed to be used this way, but I agree we should try to be clever here. Since this is an optimization I would track it with a dedicated issue and a separate PR ?